server/src/main/java/org/elasticsearch/indices/recovery/MultiFileTransfer.java

```diff
@@ -0,0 +1,211 @@
+            // These exceptions will be ignored as we record only the first failure, log them for debugging purpose.
+            items.stream().filter(item -> item.v1().failure != null).forEach(item ->
+                logger.debug(new ParameterizedMessage("failed to transfer a file chunk request {}", item.v1().md), item.v1().failure));
+            return;
+        }
+        try {
+            for (Tuple<FileChunkResponseItem, Consumer<Exception>> item : items) {
+                final FileChunkResponseItem resp = item.v1();
+                if (resp.requestSeqId == UNASSIGNED_SEQ_NO) {
+                    continue; // not an actual item
+                }
+                requestSeqIdTracker.markSeqNoAsProcessed(resp.requestSeqId);
+                if (resp.failure != null) {
+                    handleError(resp.md, resp.failure);
+                    throw resp.failure;
+                }
+            }
+            while (requestSeqIdTracker.getMaxSeqNo() - requestSeqIdTracker.getProcessedCheckpoint() < maxConcurrentFileChunks) {
+                final Tuple<StoreFileMetaData, Request> request = readAheadRequest != null ? readAheadRequest : getNextRequest();
+                readAheadRequest = null;
+                if (request == null) {
+                    assert currentFile == null && remainingFiles.hasNext() == false;
+                    if (requestSeqIdTracker.getMaxSeqNo() == requestSeqIdTracker.getProcessedCheckpoint()) {
+                        onCompleted(null);
+                    }
+                    return;
+                }
+                final long requestSeqId = requestSeqIdTracker.generateSeqNo();
+                sendChunkRequest(request.v2(), ActionListener.wrap(
+                    r -> addItem(requestSeqId, request.v1(), null),
+                    e -> addItem(requestSeqId, request.v1(), e)));
+            }
+            // While we are waiting for the responses, we can prepare the next request in advance
+            // so we can send it immediately when the responses arrive to reduce the transfer time.
+            if (readAheadRequest == null) {
+                readAheadRequest = getNextRequest();
+            }
+        } catch (Exception e) {
+            onCompleted(e);
+        }
+    }
+
+    private void onCompleted(Exception failure) {
+        assert status == Status.PROCESSING : "status [" + status + "] failure [" + failure + "]";
+        status = failure == null ? Status.SUCCESS : Status.FAILED;
+        try {
+            try {
+                IOUtils.close(failure, this);
+            } catch (Exception e) {
+                listener.onFailure(e);
+                return;
+            }
+            listener.onResponse(null);
+        } catch (Exception ignored) {
+            // we can safely ignore this exception as it happens after we have released the resource and notified the caller.
```

Comment:
why do we need this?

Reply:
++ I don't think we should have this. If we run into a `Exception` in the `listener.onResponse` we should fix the listener to handle that exception? Otherwise we're adding another situation where a listener behaves unexpectedly and we won't even see a log line for it?

Reply:
> we should fix the listener to handle that exception

++. Fixed in https://github.com/elastic/elasticsearch/pull/44040/commits/6d768948069f004c67d98b4e050e4811018605dc.